File: include/blocks/Battery.c
#include <stdio.h>
#include <string.h>
#include <utils.h>

static unsigned short batindex = 0;
static char batsym[][5] = {" ", " ", " ", " ", " "};

int blk_bat(char *value) {
  int capacity;
  char bat[256], status[16], batfilepath[64], *sym;
  unsigned batno = 0, batsymlen = sizeof(batsym) / sizeof(batsym[0]);

  int count = 0;
  while (1) {
    sprintf(batfilepath, "/sys/class/power_supply/BAT%d/capacity", batno);
    if (fpscanf(batfilepath, "%d", &capacity) == -1)
      break;

    sprintf(batfilepath, "/sys/class/power_supply/BAT%d/status", batno++);
    if (fpscanf(batfilepath, "%s", status) == -1)
      break;

    sym = strcmp(status, "Charging") == 0 ? batsym[batindex]
          : strcmp(status, "Discharging") == 0
              ? batsym[capacity * batsymlen / 100]
              : batsym[batsymlen - 1];
    sprintf(bat, count++ == 0 ? "%s %3d" : "| %s %3d", sym, capacity);
  }

  if (count == 0)
    return -1;

  batindex = (batindex + 1) % batsymlen;
  return sprintf(value, "%s", bat);
}

File: src/blocks/battery.rs
use super::{Block, ValueRunner};
use std::{fs::read_to_string, sync::Arc};

const BAT_SYM_LEN: usize = 5;

struct BatteryRunner {
    batindex: usize,
    batsym: [&'static str; BAT_SYM_LEN],
}

impl ValueRunner for BatteryRunner {
    fn fmt_value(&mut self, string: String) -> String {
        format!(" {}  ", string)
    }

    fn get_value(&mut self) -> Option<String> {
        let capacity_str =
            read_to_string("/sys/class/power_supply/BAT0/capacity").ok()?;
        let capacity: usize = capacity_str.lines().next()?.parse().ok()?;

        let status_str =
            read_to_string("/sys/class/power_supply/BAT0/status").ok()?;

        let status = status_str.lines().next()?;

        let sym = match status {
            "Charging" => self.batsym[self.batindex],
            "Discharging" => self.batsym[capacity * BAT_SYM_LEN / 100],
            _ => self.batsym.iter().last()?,
        };

        self.batindex = (self.batindex + 1) % BAT_SYM_LEN;
        Some(format!("{} {:3}%", sym, capacity))
    }
}

pub fn create_battery_blk() -> Block {
    Block::Value {
        default_value: "battery: ?",
        interval_ms: 1000,
        create_runner: Arc::new(|| {
            Box::new(BatteryRunner {
                batindex: 0,
                batsym: [" ", " ", " ", " ", " "],
            })
        }),
    }
}
